### PR TITLE
ORTModule memory improvement

### PR DESCRIPTION
**Pull request by pengwa**
## Dependency

https://github.com/microsoft/onnxruntime/pull/19007

## ORTModule memory efficient gradient management

Previously I have tried to solve the coarsed-grained gradient accumulation/update problem in ORTModule with https://github.com/microsoft/onnxruntime/pull/8979, while that resolution somehow is not fully validated with DDP or there is user hooks on the gradient accumulation on torch parameter.

This PR is addressing the problem in the similar approach as PR 8979, e.g. trigger gradient accumulation once ORT computed the grad, but instead of use a AccumulateGrad op, this time with a ONNX operator PythonOp, internally it will call param.backward(grad), which will help handle all related hooks correctly.

## Design

Check the details from

https://microsoftapc-my.sharepoint.com/:p:/g/personal/pengwa_microsoft_com/EaaBq4EzsFhOmsDEXCG7Ba4Bb9bwd0O2sFV_JXJ4jBLYLA?e=7Sz2g8&nav=eyJzSWQiOjI3MSwiY0lkIjozMjE4NzI1NDIzfQ

## Convergence Validation:

image

differences are on mostly 0.000x, sometimes 0.00x, which may comes from the different order gradient apply happens before or after this change (on deepspeed zero stage 2)

## TODO

Consolidate the logic with Stage3's similar logic.

### CODE REVIEW

**Review by snnn — APPROVED** [review]
The yaml file change looks good to me. I didn't look the other parts.

File: orttraining/orttraining/python/training/ortmodule/_inference_manager.py [hunk 1]
```diff
@@ -162,8 +162,10 @@ def forward(self, *inputs, **kwargs):
             if self._runtime_options.enable_zero_stage3_support:
                 self._append_pull_weight_trigger_as_input(kwargs, self._device)
 
+            param_to_append_as_onnx_graph_inputs = self._graph_initializers
+
             prepared_input_list, _, _ = _io._combine_input_buffers_initializers(
-                self._graph_initializers,
+                param_to_append_as_onnx_graph_inputs,
```
Comment (askhade) [inline orttraining/orttraining/python/training/ortmodule/_inference_manager.py:168]: I dont understand this change in inference_manager.py

**Review by askhade — APPROVED** [review]

**Review by snnn — APPROVED** [review]

**Comment by pengwa** [comment]
Thank you @askhade, @snnn!

File: orttraining/orttraining/python/training/ortmodule/_mem_efficient_grad_mgmt.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    named_params: dict[str, torch.nn.parameter.Parameter],
+) -> tuple[bool, ModelProto]:
+    """This function is used to enable zero stage3 compatibility.
+
+    Args:
+        exported_model (ModelProto): The exported model.
+        named_params (Optional[Dict[str, torch.nn.parameter.Parameter]]): The full parameter map.
+
+    Returns:
+        tuple[bool, ModelProto]: A tuple of bool and ModelProto. The bool indicates whether the model is modified.
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/ortmodule/_mem_efficient_grad_mgmt.py:49]: For future reference: no need to include type information in docstrings as they are in the function signature already.
